dids/diddht/internal/dns/dns.go

```diff
@@ -0,0 +1,199 @@
+	for name, data := range rec.records {
+		switch {
+		case strings.HasPrefix(name, "_k"):
+			var vMethod didcore.VerificationMethod
+			if err := UnmarshalVerificationMethod(data, &vMethod); err != nil {
+				// TODO handle error
+				continue
+			}
+
+			// TODO somehow we need to keep track of the order - verification method index should keep entryID order
+			// extracting kN from _kN._did
+			entryID := strings.Split(name, ".")[0][1:]
+			relationships, ok := relationshipMap[entryID]
+
+			if !ok {
+				// no relationships
+				continue
+			}
+
+			opts := []didcore.Purpose{}
+			for _, r := range relationships {
+				if o, ok := vmPurposeDNStoDID[r]; ok {
+					opts = append(opts, o)
+				}
+			}
+
+			document.AddVerificationMethod(
+				vMethod,
+				didcore.Purposes(opts...),
+			)
+		case strings.HasPrefix(name, "_s"):
+			var s didcore.Service
+			if err := UnmarshalService(data, &s); err != nil {
+				// TODO handle error
+				continue
+			}
+			document.AddService(&s)
+		case strings.HasPrefix(name, "_cnt"):
+			// TODO add controller https://did-dht.com/#controller
+			// optional field
+			// comma-separated list of controller DID identifiers
+			document.Controller = strings.Split(data, ",")
+		case strings.HasPrefix(name, "_aka"):
+			// TODO add aka https://did-dht.com/#also-known-as
+			document.AlsoKnownAs = strings.Split(data, ",")
+		default:
+		}
+	}
+
+	return document, nil
+}
+
+// parseDNSDID takes the bytes of the DNS representation of a DID and creates an internal representation
+// used to create a DID document
+// TODO move this in it's own internal package
+func parseDNSDID(data []byte) (*decoder, error) {
+	var p dnsmessage.Parser
+	if _, err := p.Start(data); err != nil {
+		return nil, err
+	}
+
+	didRecord := decoder{
+		records: make(map[string]string),
+	}
+
+	// need to skip questions to move the index to the right place to read answers
+	if err := p.SkipAllQuestions(); err != nil {
+		return nil, err
+	}
+
+	for {
+		h, err := p.AnswerHeader()
+		if errors.Is(err, dnsmessage.ErrSectionDone) {
+			break
+		}
+
+		if h.Type != dnsmessage.TypeTXT {
+			continue
+		}
+
+		value, err := p.TXTResource()
+		if err != nil {
+			// TODO check what kind of error and see if this should fail
+			return nil, err
+		}
+
+		name := h.Name.String()
+		fullTxtRecord := strings.Join(value.TXT, "")
+		if strings.HasPrefix(name, "_did") {
+			didRecord.id = strings.TrimSuffix(strings.TrimPrefix(name, "_did."), ".")
+			didRecord.rootRecord = fullTxtRecord
+			continue
+		}
+
+		didRecord.records[h.Name.String()] = fullTxtRecord
+	}
+
+	return &didRecord, nil
+}
+
+// TODO on the diddhtrecord we should validate the minimum reqs for a valid did
+func parseVerificationRelationships(rootRecord string) (map[string][]string, error) {
+	rootRecordProps, err := parseTXTRecordData(rootRecord)
+	if err != nil {
+		return nil, err
+	}
+	// reverse the map to get the relationships
+	relationshipMap := map[string][]string{}
```

Comment:
nit here and elsewhere: https://github.com/uber-go/guide/blob/master/style.md#initializing-maps

Reply:
whoops. I should know this 🤦